Docs 2807 New "Contributing" article: Software Translation (#1490)
truenas/documentation: content/Contributing/UITranslations.md

---

> title: "String Translations"

Reviewer: ```suggestion
title: "Web Interface Translations"
```

> description: "How to contribute CORE text string translations."

Reviewer: Is this article meant to be applicable for both CORE and SCALE? If so, then I have a small suggestion.
```suggestion
description: "How to contribute TrueNAS interface text string translations."
```

weight: 5
tags:
- coretranslate
- scaletranslate
---

## How to Contribute CORE and SCALE Text String Translations

> {{< hint info >}}
> Users can contribute text string translations for the TrueNAS translation project.  
> The first step is to sign into GitHub. Navigate to the TrueNAS webui directory, the main page of the repository.
> {{< /hint >}}

Reviewer: This doesn't need to be in a hint box when it is the only content in the section and isn't really an aside. The suggestion takes the text here and reworks this as a bit of introductory content that "sells" the idea, lists the requirements, and offers a tldr; for the step-by-step instructions.

```suggestion
By default, TrueNAS is available in English.
However, users can contribute text string translations for the TrueNAS web interface to help make TrueNAS available in many other languages!
Contributing requires a GitHub account and the ability to translate English text into other languages.

To translate the TrueNAS web interface, sign into GitHub and go to the TrueNAS **webui** repository. Find the appropriate translation file, edit it, and send a Pull Request with your translations to the project.
For step-by-step instructions, see the rest of this article.
```

> ### Navigate to the Directory

Reviewer: ```suggestion
### Go to the Localization Directory
```

> Navigate to this location: [TrueNAS translation directory](https://github.com/truenas/webui/tree/master/src/assets/i18n).  

Reviewer: Minor suggestion, but Go is a simpler and easier to digest word than Navigate.
```suggestion
Go to this location: [TrueNAS translation directory](https://github.com/truenas/webui/tree/master/src/assets/i18n).  
```

> The .json language conversion files are located within the "webui/src/assets/i18n" directory.

Reviewer: Quotes are not preferred due to their getting accidentally pulled into highlight/copy/paste operations so frequently. Bold is a good way to emphasize something specific in a way that allows easy copy/paste.
```suggestion
The .json language conversion files are located within the **webui/src/assets/i18n** directory.
```

![GitHubWebUIDirectory](/images/Contribute/GitHubWebUIDirectory.png "GitHub TrueNAS webui directory")

### Select the Translation File

> Choose the language .json file that represents the translation language.  

Reviewer: ```suggestion
Choose the .json file that represents the specific translation language.
```

More information about 2 letter ISO 639-1 codes located at: [ISO 639-1 Codes](https://www.loc.gov/standards/iso639-2/php/code_list.php).

Click the .json filename. The file opens in GitHub's web-based file editor. Click the pencil icon to begin editing the file.  

![GitHubWebUITranslationFile](/images/Contribute/GitHubWebUITranslationFile.png "TrueNAS language translation file")

### Edit the Article

Work from the last line of code in the file back up to the beginning.
Start at the last line of the file, and check for a translation. Some files have no translations listed. In this instance, begin with the last line of the file.  

> If the last line of the file contains a translation, scroll up (descending numerical order). You are looking to the next available line of code for which there is no language translation listed. Begin translating at this point.

Reviewer: ```suggestion
If the last line of the file contains a translation, scroll up (descending numerical order). Look for the next available line of code for which there is no language translation listed. Begin translating at this point.
```

> Some strings may use [ICU Message Format](https://formatjs.io/docs/core-concepts/icu-syntax/#plural-format) for pluralization. You can move tokens around and adjust them for your language.

Reviewer: ```suggestion
Some strings can use [ICU Message Format](https://formatjs.io/docs/core-concepts/icu-syntax/#plural-format) for pluralization. You can move tokens around and adjust them for your language.
```

![GitHubWebUIEditTranslationFile](/images/Contribute/GitHubWebUIEditTranslationFile.png "TrueNAS edit language translation file")

### Translate the Text Strings

 The string in English that needs translation appears first, enclosed with double quotes. After that there is a colon, followed by a second pair of double quotes. Enter your translation within the second pair of double quotes. Click on the **Preview changes** tab to view your changes as you make them. If you don't want to save your changes click **Cancel changes**.  

![GitHubWebUITranslationEntry](/images/Contribute/GitHubWebUITranslationEntry.png "TrueNAS language translation snippet")

### Begin a New Pull Request

>  Click **Propose changes** to save the file. This will create a new fork of the repository. It will also open a pull request. A pull request sends the translated text strings to TrueNAS for review. Once approved, these translations get merged into the master branch.

Reviewer: tense agreement
```suggestion
 Click **Propose changes** to save the file. This creates a new fork of the repository. It also opens a pull request. A pull request sends the translated text strings to TrueNAS for review. Once approved, these translations get merged into the master branch.
```

> Before submitting the pull request, please consider the following:

Reviewer: ```suggestion
Before submitting the pull request, please consider:
```
